buildifier/utils/diagnostics.go

```diff
@@ -0,0 +1,137 @@
+package utils
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/bazelbuild/buildtools/build"
+	"github.com/bazelbuild/buildtools/warn"
+	"sort"
+	"strings"
+)
+
+// Diagnostics contains diagnostic information returned by formatter and linter
+type Diagnostics struct {
+	Success bool               `json:"success"` // overall success (whether all files are formatted properly and have no warnings)
+	Files   []*FileDiagnostics `json:"files"`   // diagnostics per file
+}
+
+func (d *Diagnostics) Format(format string, verbose bool) string {
+	switch format {
+	case "text", "":
+		var output strings.Builder
+		for _, f := range d.Files {
+			for _, w := range f.Warnings {
+				formatString := "%s:%d: %s: %s (%s)\n"
+				if !w.Actionable {
+					formatString = "%s:%d: %s: %s [%s]\n"
+				}
+				output.WriteString(fmt.Sprintf(formatString,
+					f.Filename,
+					w.Start.Line,
+					w.Category,
+					w.Message,
+					w.URL))
+			}
+			if !f.Formatted {
+				rewrites := []string{}
+				for category, count := range f.Rewrites {
+					if count > 0 {
+						rewrites = append(rewrites, category)
+					}
+				}
+				log := ""
+				if len(rewrites) > 0 {
+					sort.Strings(rewrites)
+					log = " " + strings.Join(rewrites, " ")
+				}
+				output.WriteString(fmt.Sprintf("%s # reformat%s\n", f.Filename, log))
+			}
+		}
+		return output.String()
+	case "json":
+		var result []byte
+		if verbose {
+			result, _ = json.MarshalIndent(*d, "", "    ")
+		} else {
+			result, _ = json.Marshal(*d)
+		}
+		return string(result) + "\n"
+	}
+	return ""
+}
+
+type FileDiagnostics struct {
+	Filename  string         `json:"filename"`
+	Formatted bool           `json:"formatted"`
```

Comment:
What is `Formatted`?
Is it set to true because the file has been reformatted? Is it set to true when the file was already formatted?

Reply:
Means the file is formatted correctly. Currently the diagnostics is only available with `--mode=check`, so can't mean "reformatted".